src/main/kotlin/org/rust/cargo/project/settings/RsProjectSettingsServiceBase.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.project.settings
+
+import com.intellij.codeInsight.daemon.DaemonCodeAnalyzer
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.BaseState
+import com.intellij.openapi.components.SimplePersistentStateComponent
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Disposer
+import com.intellij.util.messages.Topic
+import org.jetbrains.annotations.TestOnly
+import org.rust.cargo.project.settings.RsProjectSettingsServiceBase.StateBase
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.memberProperties
+
+abstract class RsProjectSettingsServiceBase<T : StateBase<T>>(val project: Project, state: T) : SimplePersistentStateComponent<T>(state) {
+
+    abstract class StateBase<T : StateBase<T>> : BaseState() {
+        abstract fun copy(): T
+    }
+
+    @Retention(AnnotationRetention.RUNTIME)
+    @Target(AnnotationTarget.PROPERTY)
+    protected annotation class AffectsCargoMetadata
+
+    @Retention(AnnotationRetention.RUNTIME)
+    @Target(AnnotationTarget.PROPERTY)
+    protected annotation class AffectsHighlighting
+
+    fun modify(action: (T) -> Unit) {
+        val oldState = state.copy()
+        val newState = state.also(action)
+        val event = createSettingsChangedEvent(oldState, newState)
+        notifySettingsChanged(event)
+    }
+
+    @TestOnly
+    fun modifyTemporary(parentDisposable: Disposable, action: (T) -> Unit) {
+        val oldState = state
+        loadState(oldState.copy().also(action))
+        Disposer.register(parentDisposable) {
+            loadState(oldState)
+        }
+    }
+
+    companion object {
+        val RUST_SETTINGS_TOPIC: Topic<RsSettingsListener> = Topic.create(
+            "rust settings changes",
+            RsSettingsListener::class.java,
+            Topic.BroadcastDirection.TO_PARENT
+        )
+    }
+
+    interface RsSettingsListener {
+        fun <T : StateBase<T>> settingsChanged(e: SettingsChangedEventBase<T>)
```

Comment:
Not sure we need to have such a complex listener, but I don't have constructive suggestions on how to improve it